// raml-javascript-generator/dist/templates/package.json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var paramCase = require("param-case");
var api_1 = require("../support/api");
function default_1(api) {
    return JSON.stringify({
        name: paramCase(api.title),
        version: '0.0.0',
        description: api.description,
        main: 'index.js',
        files: [
            'index.js'
        ],
        repository: {
            type: 'git',
            url: 'git://github.com/mulesoft/raml-javascript-generator.git'
        },
        keywords: [
            'raml-api'
        ],
        author: 'MuleSoft, Inc.',
        license: 'Apache-2.0',
        bugs: {
            url: 'https://github.com/mulesoft/raml-javascript-generator/issues'
        },
        homepage: 'https://github.com/mulesoft/raml-javascript-generator',
        dependencies: {
            'client-oauth2': api_1.hasSecurity(api, 'OAuth 2.0') ? '^2.1.0' : undefined,
            'xtend': '^4.0.1',
            'popsicle': '^5.0.0',
            'setprototypeof': '^1.0.1'
        }
    }, null, 2) + '\n';
}
exports.default = default_1;
